src/core/adapters/personalization.ts

```diff
@@ -0,0 +1,126 @@
+import Actions from '../actions';
+import Configuration from '../configuration';
+import ConfigAdapter from './configuration';
+import Selectors from '../selectors';
+import Store from '../store';
+
+namespace Personalization {
+  type ExtractableAction = Actions.SelectRefinement & Actions.AddRefinement;
+
+  export const DAYS_IN_SECONDS = 86400;
+
+  export const extractBias = (action: ExtractableAction, state: Store.State) => {
+    const config = Selectors.config(state).personalization.realTimeBiasing;
+    const byId = Selectors.realTimeBiasesById(state);
+    const { field, value } = extractRefinement(action, state);
+
+    if (!config.attributes[field]) {
+      return null;
+    }
+
+    return {
+      variant: field,
+      key: value,
+      bias: (byId[field] && byId[field][value]) ? {
+        ...byId[field][value],
+        lastUsed: Math.floor(Date.now() / 1000)
+      } : generateNewBias()
+    };
+  };
+
+  // tslint:disable-next-line max-line-length
+  export const extractRefinement = ({ type, payload }: ExtractableAction, state: Store.State): { field: string, value: string } => {
+    // tslint:disable-next-line one-variable-per-declaration
+    let field = undefined, value = undefined;
+
+    switch (type) {
+      case Actions.ADD_REFINEMENT:
+        if (payload.range) {
+          break;
+        }
+        field = payload.navigationId;
+        value = payload.value;
+        break;
+      case Actions.SELECT_REFINEMENT:
+        const refinement = Selectors.refinementCrumb(state, payload.navigationId, payload.index);
+        field = refinement.field;
+        value = refinement.value;
+        break;
+      default:
+    }
+
+    return { field, value };
+  };
+
+  export const generateNewBias = () => ({
+    lastUsed: Math.floor(Date.now() / 1000)
+  });
+
+  // tslint:disable-next-line max-line-length
+  export const transformToBrowser = (state: Store.Personalization.Biasing, reducerKey: string): BrowserStorageState => ({
+    allIds: state.allIds ? state.allIds.map(({ variant, key }) => ({
+      variant,
+      key,
+      ...state.byId[variant][key]
```

Comment:
do we care that the values of `variant` or `key` might be overridden by this spread?

Reply:
No, that specific byId should not be storing variant or key, as it is already indexed by that data.